frontend/src/editor/indent.ts

```diff
@@ -0,0 +1,189 @@
+  tr: Transaction,
+  pos: number,
+  delta: number,
+  options: IndentOptions,
+): Transaction => {
+  const node = tr.doc?.nodeAt(pos)
+  if (!node) return tr
+
+  const indent = clamp(
+    (node.attrs.indent || 0) + delta,
+    options.minIndentLevel * options.indentRange,
+    options.maxIndentLevel * options.indentRange,
+  )
+  if (indent === node.attrs.indent) return tr
+
+  return tr.setNodeMarkup(pos, node.type, { ...node.attrs, indent }, node.marks)
+}
+
+const getIndent =
+  (): KeyboardShortcutCommand =>
+  ({ editor }) => {
+    return editor.can().sinkListItem("listItem")
+      ? editor.chain().focus().sinkListItem("listItem").run()
+      : editor.chain().focus().indent().run()
+  }
+
+const getOutdent =
+  (outdentOnlyAtHead: boolean): KeyboardShortcutCommand =>
+  ({ editor }) => {
+    if (outdentOnlyAtHead && editor.state.selection.$head.parentOffset > 0)
+      return false
+
+    return editor.can().liftListItem("listItem")
+      ? editor.chain().focus().liftListItem("listItem").run()
+      : editor.chain().focus().outdent().run()
+  }
+
+export const Indent = Extension.create<IndentOptions, never>({
+  name: "indent",
+
+  addOptions() {
+    return {
+      names: ["heading", "paragraph"],
+      indentRange: 40,
+      minIndentLevel: 0,
+      maxIndentLevel: 10,
+      defaultIndentLevel: 0,
+      HTMLAttributes: {},
+    }
+  },
+
+  addGlobalAttributes() {
+    return [
+      {
+        types: this.options.names,
+        attributes: {
+          indent: {
+            default: this.options.defaultIndentLevel,
+            renderHTML: (attributes) => ({
+              style: `margin-left: ${attributes.indent}px!important;`,
+            }),
+            parseHTML: (element) =>
+              parseInt(element.style.marginLeft, 10) ||
+              this.options.defaultIndentLevel,
+          },
+        },
+      },
+    ]
+  },
+
+  addCommands() {
+    return {
+      indent:
+        () =>
+        ({ tr, state, dispatch, editor }: CommandProps) => {
+          tr = updateIndentLevel(
+            tr.setSelection(state.selection),
+            this.options,
+            editor.extensionManager.extensions,
+            "indent",
+          )
+          if (tr.docChanged && dispatch) {
+            dispatch(tr)
+            return true
+          }
+          return false
+        },
+      outdent:
+        () =>
+        ({ tr, state, dispatch, editor }: CommandProps) => {
+          tr = updateIndentLevel(
+            tr.setSelection(state.selection),
+            this.options,
+            editor.extensionManager.extensions,
+            "outdent",
+          )
+          if (tr.docChanged && dispatch) {
+            dispatch(tr)
+            return true
+          }
+          return false
+        },
+    }
+  },
+
+  addKeyboardShortcuts() {
+    return {
+      Tab: getIndent(),
+      "Shift-Tab": getOutdent(false),
+      Backspace: getOutdent(true),
+      "Mod-]": getIndent(),
```

Comment:
I thought we have just the tab command... What kind of command is this? How can I trigger it?